src/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -146,28 +140,20 @@ public void NextProperty()
         {
             EndProperty();
 
-            if (PropertyEnumeratorActive)
+            int maxPropertyIndex = JsonClassInfo.PropertyCacheArray.Length;
+
+            ++PropertyEnumeratorIndex;
+            if (PropertyEnumeratorIndex >= maxPropertyIndex)
             {
-                int len = JsonClassInfo.PropertyCacheArray.Length;
-                if (PropertyEnumeratorIndex < len)
+                if (PropertyEnumeratorIndex > maxPropertyIndex)
                 {
-                    if ((PropertyEnumeratorIndex == len - 1) && JsonClassInfo.DataExtensionProperty != null)
-                    {
-                        ExtensionDataStatus = ExtensionDataWriteStatus.Writing;
-                    }
-
-                    PropertyEnumeratorIndex++;
-                    PropertyEnumeratorActive = true;
+                    ExtensionDataStatus = ExtensionDataWriteStatus.Finished;
                 }
-                else
+                else if (JsonClassInfo.DataExtensionProperty != null)
```

Comment:
Are all these branches covered with tests?

Reply:
Yes the first subbranch is hit every time we don't have an extension property and the second subbranch is hit every time we have an extension property.